File: .idea/app/src/main/java/com/chevstrap/rbx/CustomWatcher.java
package com.chevstrap.rbx;

import com.chevstrap.rbx.Integrations.ActivityWatcher;
import com.chevstrap.rbx.Models.Entities.ActivityData;

import java.util.Collection;

public class CustomWatcher {
    private static final String LOG_IDENT = "CustomWatcher";
    private static CustomWatcher instance;
    private ActivityWatcher activityWatcher;
    private ActivityData activityData;

    public static synchronized CustomWatcher getInstance() {
        if (instance == null) {
            instance = new CustomWatcher();
        } else {
            App.getLogger().writeLine(LOG_IDENT, "CustomWatcher instance already exists");
        }
        return instance;
    }

    private CustomWatcher() {
        try {
            activityWatcher = new ActivityWatcher();
        } catch (Exception e) {
            App.getLogger().writeException(LOG_IDENT, e);
        }
    }

    public void setActivityData(ActivityData data) {
        this.activityData = data;
    }

    public ActivityData getActivityData() {
        return activityData;
    }

    public void run() {
        if (activityWatcher != null) {
            activityWatcher.runWatcher();
        }
    }

    public Collection<ActivityData> getAllHistoryServer() {
        if (activityWatcher != null) {
            return activityWatcher.history;
        }
        return null;
    }

    public void Dispose() {
        App.getLogger().writeLine(LOG_IDENT, "Disposing CustomWatcher");

        if (activityWatcher != null && !activityWatcher.isStopMonitoring()) {
            activityWatcher.Dispose();
        }

        activityWatcher = null;
        activityData = null;
        instance = null;
    }
}

File: app/src/main/java/com/chevstrap/rbx/CustomWatcher.kt
package com.chevstrap.rbx

import com.chevstrap.rbx.integrations.ActivityWatcher
import com.chevstrap.rbx.integrations.RobloxDiscordRPC
import com.chevstrap.rbx.models.entities.ActivityData

class CustomWatcher private constructor() {

    @JvmField
    var activityData: ActivityData? = null

    var activityWatcher: ActivityWatcher? = null
        private set

    var robloxDiscordRPC: RobloxDiscordRPC? = null
        private set

    var isDisposed: Boolean = false
        private set

    private var running = false

    init {
        createWatcher()
    }

    private fun createWatcher() {
        try {
            val watcher = ActivityWatcher()
            activityWatcher = watcher
            robloxDiscordRPC = RobloxDiscordRPC(watcher)
        } catch (ex: Exception) {
            App.logger.writeException(LOG_IDENTIFIER, ex)
            activityWatcher = null
            robloxDiscordRPC = null
        }
    }

    @Synchronized
    fun dispose() {
        if (isDisposed) return

        isDisposed = true
        running = false

        App.logger.writeLine(LOG_IDENTIFIER, "Disposing CustomWatcher")

        val watcher = activityWatcher
        activityWatcher = null
        robloxDiscordRPC = null
        activityData = null

        try {
            if (watcher != null && !watcher.isStopMonitoring) {
                watcher.dispose()
            }
        } catch (ex: Exception) {
            App.logger.writeException(LOG_IDENTIFIER, ex)
        }

        synchronized(CustomWatcher::class.java) {
            if (instance === this) {
                instance = null
            }
        }
    }

    @Synchronized
    fun run() {
        if (isDisposed || running) return

        try {
            activityWatcher?.let {
                if (!it.isStopMonitoring) {
                    it.dispose()
                }
            }

            activityWatcher = null
            robloxDiscordRPC = null

            val watcher = ActivityWatcher()
            val rpc = RobloxDiscordRPC(watcher)

            activityWatcher = watcher
            robloxDiscordRPC = rpc

            running = true
            watcher.runWatcher()
        } catch (ex: Exception) {
            running = false
            activityWatcher = null
            robloxDiscordRPC = null
            App.logger.writeException(LOG_IDENTIFIER, ex)
        }
    }

     val allHistoryServer: MutableCollection<ActivityData>
        get() = activityWatcher?.history
            ?.filterNotNull()
            ?.toMutableList()
            ?: mutableListOf()

    companion object {
        private const val LOG_IDENTIFIER = "CustomWatcher"

        private var instance: CustomWatcher? = null

        @JvmStatic
        fun getInstance(): CustomWatcher {
            synchronized(CustomWatcher::class.java) {
                return instance ?: CustomWatcher().also {
                    instance = it
                }
            }
        }
    }
}
